Centralize path handler logic in controller server

mini-1235 opened the pull request:
This PR continues the work in #4789 
Fixes #4757 
Fixes #4304 
> This is a continuation of work started by @ottojo in draft PR #4593
> 
> I'll quote their introduction and, expand a bit afterwards, and update any fields that have changed.
> 
> > The controller server uses a configurable goal checker to determine if the robot has completed its current path as defined by the global plan. Current goal checkers compare the goal (end of path) pose and twist to the current robot state, but that is not sufficient in all cases.
> > In some use cases, the path will visit the goal multiple times, and the goal might coincide with the starting position. It is still desired that the robot follows the entire path, instead of immediately ending navigation once the goal pose is reached.
> > This PR adds a parameter to the GoalChecker interface to inform the goal checker of the current path, which enables building more sophisticated goal checkers that (for example) take progress along the path into account.
> > We already use such a goal checker internally, which just subscribes to the global plan via the appropriate ROS topic, but i think this here is the cleaner solution (and also avoids race conditions of checking against an old path in the goal checker...)
> > I don't have a nice testing setup and a goal checker using this interface yet (other than our own robot and the mentioned goal checker), which is why I mark this as draft for now.
> 
> I've cherry-picked @ottojo 's commit on top of the latest main branch, then added a new basic goal checker PathCompleteGoalChecker, which is a subclass of SimpleGoalChecker, but has an additional parameter (path_length_tolerance, default=1), and checks that the current path has <= path_length_tolerance waypoints before allowing the normal SimpleGoalChecker completion logic to take its course.
> 
> The result is a trivial extension of SimpleGoalChecker that should be immune to the current_pose = goal_pose short-circuit problem/bug. Rather than try to create an elaborate unit test that navigates a course, I stuck to first principles to ensure that the new plugin behaves the same as SimpleGoalChecker with <= path_length_tolerance waypoints, and that it does not complete with > path_length_tolerance waypoints.
> 
> ## Basic Info
> Info	Please fill out this column
> Ticket(s) this addresses	#4304
> Primary OS tested on	Ubuntu
> Robotic platform tested on	(WIP)
> Does this PR contain AI generated software?	No
> ## Description of contribution in a few bullet points
> * [x]   Extend `GoalChecker::isGoalReached` interface with argument for current path
> * [x]   Add this argument to the in-tree goal checker implementations
> * [x]   Add a goal checker which checks for path-completion
> * []  Define a test-scenario with global plan that returns to start pose
> 
> ## Description of documentation updates required from your changes
> * [ ]   Migration guide for out-of-tree goal checkers
> 
> ## Future work that may be required in bullet points
> * Implementing more sophisticated goal checkers
> 
> #### For Maintainers:
> * [ ]   Check that any new parameters added are updated in docs.nav2.org
> * [ ]   Check that any significant change is added to the migration guide
> * [ ]   Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
> * [ ]   Check that any new functions have Doxygen added
> * [ ]   Check that any new features have test coverage
> * [ ]   Check that any new plugins is added to the plugins page
> * [ ]   If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

mini-1235 commented:
I have spent some time reading into MPPI's path handler, I can see there are some improvements like https://github.com/ros-navigation/navigation2/pull/3659, https://github.com/ros-navigation/navigation2/pull/3443. Also, there is a related PR for RPP in #4763, maybe we need to check if these changes are applicable to all controllers? If yes, I guess it will be easier to move them in Controller Server

I also noticed that the DWB/MPPI need the path from planner to find its goal, should we pass this plan via the setplan api or a new argument in computevelocitycommands?

mamariomiamo commented:
> I have spent some time reading into MPPI's path handler, I can see there are some improvements like #3659, #3443. Also, there is a related PR for RPP in #4763, maybe we need to check if these changes are applicable to all controllers? If yes, I guess it will be easier to move them in Controller Server
> 
> I also noticed that the DWB/MPPI need the path from planner to find its goal, should we pass this plan via the setplan api or a new argument in computevelocitycommands?

I think path_handler implementation from MPPI is more comprehensive as it also considers path up to inversion point. This would also be helpful when user uses Hybrid A* with reeds-shepp as the global planner.

SteveMacenski commented on nav2_controller/include/nav2_controller/controller_server.hpp:
```diff
@@ -282,6 +283,11 @@ class ControllerServer : public nav2::LifecycleNode
   // Current path container
   nav_msgs::msg::Path current_path_;
+  nav_msgs::msg::Path local_path_;
+  bool interpolate_curvature_after_goal_;
```
Is this really a controller-server specific parameter? This feels specific to a given control algorithm plugin

mini-1235 replied:
I think path handler needs to, this is used as the `reject_unit_path` in rpp's path handler 
https://github.com/ros-navigation/navigation2/blob/3c081db0cf7508451cf10651be79b4566729a40d/nav2_regulated_pure_pursuit_controller/src/path_handler.cpp#L52

SteveMacenski commented on nav2_controller/include/nav2_controller/controller_server.hpp:
```diff
@@ -282,6 +283,11 @@ class ControllerServer : public nav2::LifecycleNode
+  bool interpolate_curvature_after_goal_;
+  double max_robot_pose_search_dist_;
+  std::unique_ptr<nav2_controller::PathHandler> path_handler_;
+  nav2::Publisher<nav_msgs::msg::Path>::SharedPtr global_path_pub_;
```
Local plan publisher?

SteveMacenski commented on nav2_controller/plugins/path_complete_goal_checker.cpp:
```diff
@@ -0,0 +1,109 @@
+  const geometry_msgs::msg::Pose & query_pose, const geometry_msgs::msg::Pose & goal_pose,
+  const geometry_msgs::msg::Twist & twist, const nav_msgs::msg::Path & path)
+{
+  if(nav2_util::geometry_utils::calculate_path_length(path) > path_length_tolerance_){
```
How are we finding the closest point on the path? Imagine that you have a situation where a path is computed one time only at the start of the task and otherwise doesn't get replanned ever or only when strictly required - so you can't rely on index zero of the path being where the robot is located

SteveMacenski replied:
Ah, is this already pruned and transformed? If so then that would work :+1:

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -39,7 +40,7 @@ ControllerServer::ControllerServer(const rclcpp::NodeOptions & options)
   goal_checker_loader_("nav2_core", "nav2_core::GoalChecker"),
   default_goal_checker_ids_{"goal_checker"},
-  default_goal_checker_types_{"nav2_controller::SimpleGoalChecker"},
+  default_goal_checker_types_{"nav2_controller::PathCompleteGoalChecker"},
```
I think by 'default' I mean that this should just be built into the simple goal checker so that this just given to everyone 'for free'

mini-1235 replied:
So I should remove this new plugin and add the `path_length_tolerance` to simple goal checker, do I understand this correctly?

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -225,6 +232,7 @@ ControllerServer::on_configure(const rclcpp_lifecycle::State & state)
 
   odom_sub_ = std::make_unique<nav2_util::OdomSmoother>(node, odom_duration, odom_topic);
   vel_publisher_ = std::make_unique<nav2_util::TwistPublisher>(node, "cmd_vel");
+  global_path_pub_ = node->create_publisher<nav_msgs::msg::Path>("received_global_plan");
```
Why this? We in other contexts publish out the path after its been pruned and converted into the odom frame for the local_path publisher so that we have a sense of what the controller is working with. Why directly publish out the global plan?

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -604,14 +615,7 @@ void ControllerServer::setPlannerPath(const nav_msgs::msg::Path & path)
     throw nav2_core::InvalidPath("Path is empty.");
   }
   controllers_[current_controller_]->setPlan(path);
```
Do we need this `setPlan` if its now being given in the compute velocity command context? We do sometimes do other things in the setPlan like reset state, but maybe then it could be refactored or done in another way?

SteveMacenski replied:
> MPPI and DWB needs to know the last pose of the plan (goal)

If the computation API has this, it can be set there now I think. 

> Reset state when new plan receives

Yes I agree. 

---

Maybe what makes sense is to change the API to be more of a notification about a new path (and the raw, full global path path can be given) and we change the function name / description. But, we don't use that path for anything that the raw, global framed, full path isn't necessary for. We use the path given in the compute velocity command API instead which is pre-transformed and pruned.

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -630,14 +634,25 @@ void ControllerServer::computeAndPublishVelocity()
+  auto transformed_plan = path_handler_->transformGlobalPlan(
+    pose, max_robot_pose_search_dist_, interpolate_curvature_after_goal_);
+  // RCLCPP_INFO(get_logger(), "compute remaining distance %lf ",nav2_util::geometry_utils::calculate_path_length(transformed_plan));
+  global_path_pub_->publish(transformed_plan);
```
I think if this is pruned, in the odom frame, and such, this is a local plan

SteveMacenski replied:
Also, make this into a unique pointer `make_unique<nav_msgs::msg::Path>(local_path_)`

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -630,14 +634,25 @@ void ControllerServer::computeAndPublishVelocity()
 
   geometry_msgs::msg::Twist twist = getThresholdedTwist(odom_sub_->getRawTwist());
 
+  auto transformed_plan = path_handler_->transformGlobalPlan(
```
```suggestion
  local_path_ = path_handler_->transformGlobalPlan(
```
?

SteveMacenski reviewed (changes requested):
Sorry it took so long!

mini-1235 commented:
> There's a number of CI failures that should also be looked at, but I'm sure you know :-)

Yes, I will continue this PR once #5496 and other future PRs targeting path handler are done

SteveMacenski commented:
OK will do! This is admittedly alot so this next round is going to be more high level. 

> ... the local goal ends up being the cusping point, and the robot may stop there once it satisfies the XY and yaw tolerances ... we could add this directly into the simple goal checker, so by default the simple goal checker is checking both the global goal and local path length

Ah ok. That makes sense for the goal checker, though probably not the other elements right? When enforcing inversions is off, its the same either way since we just prune to the prune distance / costmap bounds. When is it enabled, then we prune to that which we want from the control algorithm's perspective, but not the goal checker. 

> Previously, the transform tolerance in the controller server was obtained from costmap->getTransformTolerance. I think it makes more sense for the server and its plugins to share the same tolerance, so I moved this to the controller level.

Why not have it still acquire it from costmap2D, just stored and used at the controller server level? I suppose I prefer less parameters to more, but I'm OK either way. 

> We have different pruned_plan_end for MPPI and RPP, 

Do both - first after the integrated distance, as long as that distance is within the costmap bounds.

mini-1235 commented on nav2_controller/src/parameter_handler.cpp:
```diff
@@ -0,0 +1,286 @@
+    node, "odom_topic", rclcpp::ParameterValue("odom"));
+  declare_parameter_if_not_declared(
+    node, "odom_duration", rclcpp::ParameterValue(0.3));
+  declare_parameter_if_not_declared(
```
Ok will do

mini-1235 commented on nav2_controller/include/nav2_controller/parameter_handler.hpp:
```diff
@@ -0,0 +1,113 @@
+   */
+  ~ParameterHandler();
+
+  std::mutex & getMutex() {return mutex_;}
```
Yes I forgot to add this, will add it in the next update

mini-1235 commented:
I will think about it and reply tomorrow

mini-1235 commented:
> Isn't that in effect what is done today?

I think we declare and get this parameter for all controller plugins, instead of using costmap's

https://github.com/ros-navigation/navigation2/blob/185c8b8ea83622eff792c8a9e6e69e9904b418b8/nav2_mppi_controller/src/path_handler.cpp#L40

https://github.com/ros-navigation/navigation2/blob/185c8b8ea83622eff792c8a9e6e69e9904b418b8/nav2_regulated_pure_pursuit_controller/src/parameter_handler.cpp#L51-L52

> A good suggestion I received today: It would be great to make the path handler itself a plugin so that other users can replace this implementation if they see fit. A nice side effect is it makes us have to think a bit more about the API interfaces for it to make sure its generally good for a broad range of purposes and include all the information for the APIs other implementations may want that we have access to.

I think this is a good idea. The path handler only needs the global path and the current pose as inputs, and its output would be the pruned and pre-transformed global plan. That plan is then passed on to `computeVelocityCommands` and `isGoalReached`. Since users can also implement their own controller plugins and goal checkers, this seems feasible. Maybe we can start with `SimplePathHandler` that handles for our controller plugins, or even define the output more generally as a pruned, transformed plan, since we can now support multiple path handlers. For example, we could have one for RPP that requires the path to be in robot base frame, and another one for MPPI that is in odom frame


>Another would be possibly changing the pruning distance to be proportional to time instead, to prune the distance set out by that time by the maximum velocity. That way it wasn't something that needed to be tuned for changing velocity limits.

I think to achieve this, we need to move `vx_max`, `vy_max` to the controller level, right?

> Finally, maybe it would be good to provide feedback about the current idx of the path we're on to use in the BT Navigator for computing ETA, path length remaining, and so forth. That way we handle the issue of path looping using the wrong index globally

So something like
```
<FollowPath path="{path}" controller_id="{selected_controller}" error_code_id="{follow_path_error_code}" error_msg="{follow_path_error_msg}" current_path_idx="{current_path_idx}/>
```
?

SteveMacenski commented:
I'd say path handler

SteveMacenski commented:
> For https://github.com/ros-navigation/navigation2/issues/4304:

Why so on DWB? If the goal checker / path given is restricted to the local time horizon, it shouldn't know anything about far away loops and so forth once a path segment leaves the local costmap bounds / search distance (like MPPI, RPP, etc). 

> However, when configured incorrectly, e.g. Navfn + Enable enforce_path_inversion in path handler, we might detect a cusp according to the condition here 

Yeah that is true. Setting `enforce_path_inversion` to false by default and making an note about that in the configuration guide and in the Nav2 yaml file would be good since that's an annoying one.

mini-1235 commented on nav2_bt_navigator/behavior_trees/navigate_w_replanning_distance.xml:
```diff
@@ -10,7 +10,7 @@
         <ComputePathToPose goal="{goal}" path="{path}" planner_id="{selected_planner}" error_code_id="{compute_path_error_code}" error_msg="{compute_path_error_msg}"/>
       </DistanceController>
-      <FollowPath path="{path}" controller_id="{selected_controller}" error_code_id="{follow_path_error_code}" error_msg="{follow_path_error_msg}"/>
+      <FollowPath path="{path}" controller_id="{selected_controller}" error_code_id="{follow_path_error_code}" error_msg="{follow_path_error_msg}" tracking_feedback="{tracking_feedback}"/>
```
Ok, I will also revert the changes related to distance calculation in Nav2Pose, and open a follow up PR after this is merged

mini-1235 commented on nav2_controller/plugins/simple_path_handler.cpp:
```diff
@@ -0,0 +1,285 @@
+double SimplePathHandler::getCostmapMaxExtent() const
+{
+  const double max_costmap_dim_meters = std::max(
+    costmap_ros_->getCostmap()->getSizeInCellsX(),
```
I am not sure about this one, Graceful / RPP use the API `getSizeInMetersX()`, while MPPI `getSizeInCellsX()`. I think `getSizeInCellsX()` is the one we want to use here, yes?

mini-1235 commented:
Oops, forgot to format files

SteveMacenski commented:
It didn't used to do that ... I wonder if the ament_uncrustify format got corrupted in that case.  it didn't used to do that and that was the profile in use.

is there anything above you want me to take a look at (except that last comment)? I wasn't sure if I was waiting or you wanted me to take a look again.

mini-1235 commented:
> It didn't used to do that ... I wonder if the ament_uncrustify format got corrupted in that case. it didn't used to do that and that was the profile in use.

I am not familiar with `uncrustify`, will take a look when I have time. For this PR, I will just update it manually.

> I wasn't sure if I was waiting or you wanted me to take a look again.

Yes, please take another look when you have time. I have updated things accordingly

SteveMacenski commented on nav2_regulated_pure_pursuit_controller/test/test_regulated_pp.cpp:
```diff
@@ -520,360 +456,6 @@ TEST(RegulatedPurePursuitTest, testDynamicParameter)
     results4);
 }
 
-class TransformGlobalPlanTest : public ::testing::Test
```
Were all the relevant tests for full coverage of the new plugin transferred?

mini-1235 replied:
Will add in next update

SteveMacenski commented on nav2_core/include/nav2_core/path_handler.hpp:
```diff
@@ -0,0 +1,74 @@
+ * path segments the robot has already traversed and transforming the remaining,
+ * relevant portion of the path into the costmap's global or base frame.
+ */
+class PathHandler
```
I meant just for the header/base class name (not the param for controller server config) -- but I'm OK keeping it

SteveMacenski commented on nav2_controller/include/nav2_controller/plugins/simple_path_handler.hpp:
```diff
@@ -0,0 +1,110 @@
+    * @return plan transformed in the robot base frame frame and iterator to the first pose of the global
+    * plan
+    */
+  virtual std::pair<nav_msgs::msg::Path, PathIterator> getGlobalPlanConsideringBoundsInCostmapFrame(
```
@mini-1235 I think this one is still an open item for the base plugin `nav2_core` header

SteveMacenski commented on nav2_rotation_shim_controller/src/nav2_rotation_shim_controller.cpp:
```diff
@@ -391,11 +396,11 @@ bool RotationShimController::isGoalChanged(const nav_msgs::msg::Path & path)
-  current_path_ = path;
-  primary_controller_->setPlan(path);
+  path_updated_ = rotate_to_heading_once_ ? isGoalChanged(raw_global_path) : true;
+  current_path_ = raw_global_path;
```
Should we use the new `computeVelocityCommands` for this instead for setting the current path?

SteveMacenski reviewed (commented):
Also check out CI, a number of tests don't have results or have failures

There are a number of open items above (make sure to expand the hidden topics since this is so long) and I commented on all of them with responses (if not closing due to completion).

mini-1235 commented on nav2_system_tests/src/waypoint_follower/test_case_launch.py:
```diff
@@ -48,15 +48,10 @@ def generate_launch_description() -> LaunchDescription:
     params_file = os.path.join(nav2_bringup_dir, 'params/nav2_params.yaml')
-
-    # Replace the `use_astar` setting on the params file
-    param_substitutions = {
```
This substitutes the parameter `prune_distance` and `forward_prune_distance` in DWB. Since we are centralizing it in controller server, I don't think we have these two parameters anymore in DWB

mini-1235 commented:
I think I have addressed most of the review comments here, but there are two items that I want to discuss.

> Should we use the new computeVelocityCommands for this instead for setting the current path?

https://github.com/ros-navigation/navigation2/blob/8dbc92910b020b19b9891849b3ee7cb92e1b3b03/nav2_rotation_shim_controller/src/nav2_rotation_shim_controller.cpp#L273-L281

I am not sure if we actually have use cases where forward_sampling_distance is large enough and need the raw global path here.
If we do need that, maybe it's better to pass the path handler pointer and implement APIs such as `getTransformedGoal`, `getPlan`, `getTransformedPlan` (?)
```cpp
geometry_msgs::msg::TwistStamped computeVelocityCommands(
    const geometry_msgs::msg::PoseStamped & pose,
    const geometry_msgs::msg::Twist & velocity,
    nav2_core::GoalChecker * goal_checker,
    nav2_core::PathHandler * path_handler) 
```


> Maybe this can be broken up into 2 functions: one that handles finding the start/end iterators of the path and another that actually does the transformation / checking costmap bounds and returning the updated path

> @mini-1235 I think this one is still an open item for the base plugin nav2_core header

I am thinking how we can benefit from this API change, especially after #5387 is merged. 

For example:
```cpp
auto [start, end] = path_handler.findPlanSegmentIterators() -> return the first and last point of the local plan in map frame
nav_msgs::msg::Path transformed_plan = path_handler.transformLocalPlan(start,end) -> transform that segment to odom frame and check if it is inside costmap
```

Then, 

```cpp
// Use robot pose and local plan in odom frame
tracking_feedback_msg->tracking_error = distance_to_path_segment(robot_pose, transformed_plan[0], transformed_plan[1]);
// global plan and start in map frame
tracking_feedback_msg->current_path_index = std::distance(global_plan.poses.begin(), start);
```

Is that your understanding as well?

mini-1235 commented:
Also, I haven’t applied the new dynamic parameters pattern to the controller plugins (goal_checker, path_handler, progress_checker), adding it in this PR would probably introduce another 1000 lines. I can open a separate PR for that either before or after this one, whichever you prefer

SteveMacenski commented on nav2_rotation_shim_controller/src/nav2_rotation_shim_controller.cpp:
```diff
@@ -255,7 +250,8 @@ geometry_msgs::msg::TwistStamped RotationShimController::computeVelocityCommands
   in_rotation_ = false;
-  auto cmd_vel = primary_controller_->computeVelocityCommands(pose, velocity, goal_checker);
+  auto cmd_vel = primary_controller_->computeVelocityCommands(pose, velocity, goal_checker,
+      transformed_global_plan, global_goal);
```
```suggestion
    transformed_global_plan, global_goal);
```

SteveMacenski reviewed (commented):
A couple more items. I have approved 73/95 files (and I think ~6 are just the yaml files for renaming the "Simple" path handler)

SteveMacenski commented on nav2_util/include/nav2_util/path_utils.hpp:
```diff
@@ -56,6 +59,88 @@ PathSearchResult distance_from_path(
+    if (
+      (hypot(oa_x, oa_y) == 0.0 &&
+      path.poses[idx - 1].pose.orientation !=
+      path.poses[idx].pose.orientation)
```
I think that could make sense to pull in the work from the other PR, yes :-)

mini-1235 commented:
🫠 It seems like the latest rolling sync from Rviz/Gz has caused the build to fail again

SteveMacenski commented on nav2_util/src/path_utils.cpp:
```diff
@@ -82,4 +82,43 @@ PathSearchResult distance_from_path(
+    source_pose.header.stamp = input_path.header.stamp;
+    source_pose.pose = input_pose.pose;
+
+    if (!nav2_util::transformPoseInTargetFrame(
```
Oh wow, I'm surprised, that's an obvious gap.

SteveMacenski commented on nav2_bringup/params/nav2_params.yaml:
```diff
@@ -96,6 +96,7 @@ controller_server:
     progress_checker_plugins: ["progress_checker"]
     goal_checker_plugins: ["general_goal_checker"] # "precise_goal_checker"
     controller_plugins: ["FollowPath"]
+    path_handler_plugins: ["path_handler"]
```
To just save a hassle in the other ticket on renaming these, we can rename this now to `PathHandler`

SteveMacenski commented on nav2_bringup/rviz/nav2_default_view.rviz:
```diff
@@ -413,18 +413,18 @@ Visualization Manager:
-          Shaft Diameter: 0.10000000149011612
-          Shaft Length: 0.10000000149011612
+          Shaft Diameter: 0.004999999888241291
+          Shaft Length: 0.019999999552965164
```
Revert?

SteveMacenski commented on nav2_controller/include/nav2_controller/parameter_handler.hpp:
```diff
@@ -0,0 +1,105 @@
+  const std::vector<std::string> default_goal_checker_types_{"nav2_controller::SimpleGoalChecker"};
+  const std::vector<std::string> default_controller_ids_{"FollowPath"};
+  const std::vector<std::string> default_controller_types_{"dwb_core::DWBLocalPlanner"};
+  const std::vector<std::string> default_path_handler_ids_{"path_handler"};
```
Ditto on PathHandler

SteveMacenski commented on nav2_controller/plugins/feasible_path_handler.cpp:
```diff
@@ -0,0 +1,291 @@
+    ".enforce_path_inversion", false);
+  enforce_path_rotation_ = node->declare_or_get_parameter(plugin_name +
+    ".enforce_path_rotation", false);
+  inversion_xy_tolerance_ = node->declare_or_get_parameter(plugin_name + ".inversion_xy_tolerance",
```
I dont know what's going on with our linting right now, but this would break it. Anytime there's a first `(` and a line break, it'll want you to move the contents of that scope to a new line, i.e. 

```
  inversion_xy_tolerance_ = node->declare_or_get_parameter(
    plugin_name + ".inversion_xy_tolerance", 0.2);
```

Fix that here and in other places now so that once that's working properly we have less errors to deal with. Something is definitely going on with the ament linters, both our CircleCI and the independently run GitHub Actions linter jobs are asleep.

SteveMacenski commented on nav2_controller/plugins/feasible_path_handler.cpp:
```diff
@@ -0,0 +1,291 @@
+{
+  global_plan_ = path;
+  global_plan_up_to_constraint_ = global_plan_;
+  if(enforce_path_inversion_ || enforce_path_rotation_) {
```
```suggestion
  if (enforce_path_inversion_ || enforce_path_rotation_) {
```

SteveMacenski commented on nav2_controller/plugins/feasible_path_handler.cpp:
```diff
@@ -0,0 +1,291 @@
+    global_plan_pose->header.stamp = global_pose_.header.stamp;
+    global_plan_pose->header.frame_id = global_plan_.header.frame_id;
+    nav2_util::transformPoseInTargetFrame(*global_plan_pose, costmap_plan_pose, *tf_,
+        costmap_ros_->getGlobalFrameID(), transform_tolerance_);
```
```suggestion
      costmap_ros_->getGlobalFrameID(), transform_tolerance_);
```

I feel like something must have changed recently in vscode. I'm getting alot of PRs from alot of people with mixed in 4 spaces that I cannot explain. Its not just you, its everyone :(

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -640,21 +655,36 @@ void ControllerServer::computeAndPublishVelocity()
+    path_handlers_[current_path_handler_]->getTransformedGoal(pose.header.stamp);
+  // Get the [start, end) iterators under map frame to be used for control.
+  auto [closest_point,
+    pruned_plan_end] = path_handlers_[current_path_handler_]->findPlanSegment(pose);
```
```suggestion
  auto [closest_point, pruned_plan_end] =
    path_handlers_[current_path_handler_]->findPlanSegment(pose);
```

Just my readability preference

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -173,26 +127,52 @@ ControllerServer::on_configure(const rclcpp_lifecycle::State & state)
+        get_logger(), "Created path handler : %s of type %s",
+        params_->path_handler_ids[i].c_str(), params_->path_handler_types[i].c_str());
+      path_handler->initialize(node, get_logger(), params_->path_handler_ids[i], costmap_ros_,
+          costmap_ros_->getTfBuffer());
```
```suggestion
        costmap_ros_->getTfBuffer());
```

SteveMacenski reviewed (changes requested):
This is good - pretty much down to just nitpicks due to the sheer size of this PR. 

I went through the entire open thread (took awhile) and found the following still open items:
* https://github.com/ros-navigation/navigation2/pull/5446#discussion_r2457513894 -- do we need the full global path really? I think we discussed this elsewhere below as well that perhaps we didn't.
* https://github.com/ros-navigation/navigation2/pull/5446#discussion_r2421377373 -- breaking up a function into 2 methods that would be virtual that could be overridden in particular need situations to make the base path handler implementation the most reusable.  


Per usual: new BT nodes, parameters, etc should have configuration guide updates and migration guide entries. For new BT nodes and plugins, add them to the Navigation Plugins table. I think you're ready to work on this now, I don't see any other large changes.

mini-1235 commented on nav2_rotation_shim_controller/src/nav2_rotation_shim_controller.cpp:
```diff
@@ -198,7 +196,7 @@ geometry_msgs::msg::TwistStamped RotationShimController::computeVelocityCommands
       }
     } catch (const std::runtime_error & e) {
-      RCLCPP_DEBUG(
+      RCLCPP_INFO(
```
I didn't mean to change this, but I find it useful to make it as INFO/WARNING/ERROR when fixing unit tests

SteveMacenski commented on nav2_bt_navigator/behavior_trees/navigate_through_poses_w_replanning_and_recovery.xml:
```diff
@@ -9,6 +9,7 @@
       <PipelineSequence name="NavigateWithReplanning">
         <ProgressCheckerSelector selected_progress_checker="{selected_progress_checker}" default_progress_checker="progress_checker" topic_name="progress_checker_selector"/>
         <GoalCheckerSelector selected_goal_checker="{selected_goal_checker}" default_goal_checker="general_goal_checker" topic_name="goal_checker_selector"/>
+        <PathHandlerSelector selected_path_handler="{selected_path_handler}" default_path_handler="PathHandler" topic_name="path_handler_selector"/>
```
Add to other XMLs with the selector plugins like the default nav2pose?

SteveMacenski replied:
There are others, but this is the other 'main' one I think you should add it to (also goal / progress checker) https://github.com/ros-navigation/navigation2/blob/main/nav2_bt_navigator/behavior_trees/navigate_to_pose_w_replanning_and_recovery.xml#L11-L12

SteveMacenski commented on nav2_util/src/path_utils.cpp:
```diff
@@ -82,4 +82,140 @@ PathSearchResult distance_from_path(
+{
+  static rclcpp::Logger logger = rclcpp::get_logger("transformPathInTargetFrame");
+
+  if(input_path.header.frame_id == target_frame) {
```
```suggestion
  if (input_path.header.frame_id == target_frame) {
```

SteveMacenski commented on nav2_util/src/path_utils.cpp:
```diff
@@ -82,4 +82,140 @@ PathSearchResult distance_from_path(
+
+    if (rotation_idx <= idx + 1 && inversion_idx <= idx + 1) {
+      break;
+    }
```
Please explain in a comment

mini-1235 replied:
If <= idx + 1, it means there cannot be a smaller index returned from the inversion/rotation check, otherwise a smaller index might exist, we just haven't checked it yet, I will add a comment

SteveMacenski commented:
>  If the forward sampling distance > prune distance, then the behavior may differ from what we currently have today.

Log an error if this happens and set the prune distance to be the sampling distance in that case to make sure that's not possible.

SteveMacenski commented:
> I don't think we could do that though. The rotation shim controller doesn't have access to the path_handler_ids, so it can't get these parameters directly. That means the only place we could do this logic would be in the controller server, but I don't think that's a clean or appropriate solution

OK! Docs in the configuration guide would do the trick. Practically speaking, I don't see this being an issue. In the Rotation Shim controller, if the path isn't at least the forward sampling distance long, we could log something https://github.com/ros-navigation/navigation2/blob/2b02d7d6ff6cba43c671cd883c67b65fb6b1d03c/nav2_rotation_shim_controller/src/nav2_rotation_shim_controller.cpp#L238-L241. We know this would be the case on the approach to the goal, but if the `current_path_.poses.back();` is not close to `global_goal`  then we know that is not the situation we're in, so we could long once a warning. We'd just need to pass that into the method / use a class member and then make sure everything is in a consistent frame to compare pose distances.

mini-1235 commented:
> OK! Docs in the configuration guide would do the trick. Practically speaking, I don't see this being an issue. In the Rotation Shim controller, if the path isn't at least the forward sampling distance long, we could log something
> 
> https://github.com/ros-navigation/navigation2/blob/2b02d7d6ff6cba43c671cd883c67b65fb6b1d03c/nav2_rotation_shim_controller/src/nav2_rotation_shim_controller.cpp#L238-L241
> 
> . We know this would be the case on the approach to the goal, but if the `current_path_.poses.back();` is not close to `global_goal` then we know that is not the situation we're in, so we could long once a warning. We'd just need to pass that into the method / use a class member and then make sure everything is in a consistent frame to compare pose distances.

Ok will add a warning + update docs

mini-1235 commented:
Just got back from the dentist, working on this right now :)

SteveMacenski commented:
> Just got back from the dentist, working on this right now :)

Damn, that dentist works _late_

SteveMacenski commented on nav2_util/src/path_utils.cpp:
```diff
@@ -82,4 +82,140 @@ PathSearchResult distance_from_path(
+  return true;
+}
+
+unsigned int findFirstPathConstraint(
```
Cant we just set `inversion_idx = idx +1;` and then `break;` out of the for loop?

tonynajjar commented:
I skimmed over the PR and generally looks good, but more valuably, I ran the PR against our tests suite and they succeeded. Obviously our tests don't cover the entirety of the changes in this PR but it's something!

mini-1235 commented:
> I skimmed over the PR and generally looks good, but more valuably, I ran the PR against our tests suite and they succeeded. Obviously our tests don't cover the entirety of the changes in this PR but it's something!

Thanks a lot for the review @tonynajjar  ❤️

tonynajjar commented:
>  I ran the PR against our tests suite

Please ignore that, something went wrong and I didn't test the right changes. I'll rectify and report back

tonynajjar commented on nav2_controller/plugins/feasible_path_handler.cpp:
```diff
@@ -0,0 +1,290 @@
+  reject_unit_path_ = node->declare_or_get_parameter(
+    plugin_name + ".reject_unit_path", false);
+  max_robot_pose_search_dist_ = node->declare_or_get_parameter(
+    plugin_name + ".max_robot_pose_search_dist", getCostmapMaxExtent());
```
@mini-1235 I don't see this param in the yaml file nor in the docs

tonynajjar commented on nav2_bringup/params/nav2_params.yaml:
```diff
@@ -115,6 +116,16 @@ controller_server:
       xy_goal_tolerance: 0.25
       yaw_goal_tolerance: 0.25
+      path_length_tolerance: 1.0
+    PathHandler:
```
@mini-1235 isn't `max_robot_pose_search_dist` missing?

tonynajjar replied:
No strong opinion, I just thought that we expose in the yaml file everything that is a configurable parameter. But if the philosophy is rather to expose parameters that are likely to be tuned then I also understand that

tonynajjar commented on nav2_controller/plugins/position_goal_checker.cpp:
```diff
@@ -58,8 +62,14 @@ void PositionGoalChecker::reset()
+    path_length_tolerance_)
+  {
+    return false;
+  }
```
out of curiosity, is this for optimization purposes only?

mini-1235 replied:
No, this was added here to solve https://github.com/ros-navigation/navigation2/issues/4304. Without it, the robot might incorrectly think that it has reached the goal, because it only checks xy / yaw tolerance

doisyg commented:
Amazing work!

tonynajjar commented:
@mini-1235 I found a bug: Try setting a param e.g. `ros2 param set /controller_server FollowPath.vx_max 0.5` **while** the follow_path action is running, you should get a deadlock.

tonynajjar commented:
> > I ran the PR against our tests suite
> 
> Please ignore that, something went wrong and I didn't test the right changes. I'll rectify and report back

Okay now I can say that aside from the deadlock issue, tests pass :)

mini-1235 commented:
> @mini-1235 I just tried your latest commit quickly and I don't think they fix the deadlocks, did it work for you?

Yes, FollowPath.vx_max should not be locked because it satisfies the condition here

https://github.com/mini-1235/navigation2/blob/7a7ae73dfd70b54e61f4e30b7938ee1701a5b26d/nav2_controller/src/parameter_handler.cpp#L188-L190

I also tried setting `failure_tolerance` when FollowPath is running, I see the warning log here

https://github.com/mini-1235/navigation2/blob/7a7ae73dfd70b54e61f4e30b7938ee1701a5b26d/nav2_controller/src/parameter_handler.cpp#L192-L197

mini-1235 commented:
Huge thanks to everybody involved in this PR! 🎉
